design-docs/improved-console.md

```diff
@@ -0,0 +1,422 @@
+
+```bash
+org.gradle.logger.showCategory=true                        # default=false
+org.gradle.logger.dateTimeFormat=yyyy-MM-dd HH:mm:ss:SSS Z # default=HH:mm:ss.SSS
+org.gradle.logger.showDateTime=true                        # default=false
+org.gradle.logger.showLogLevel=true                        # default=true
+```
+
+When rendering `RenderableOutputEvent`s, user configuration is taken into account.
+
+### Open issues
+* This should take source (build operation) into consideration
+
+## Story: Display overall build progress as a progress bar
+Render ProgressEvents with build completeness information as a progress bar.
+ 
+Intended to give the user a very fast way of telling whether the build will be finished soon or not.
+The work-in-progress (yellow) section shows how much of the build would be complete if all
+Operations displayed in the progressArea completed.
+
+### User-visible Changes
+We can visually represent complete and un-started tasks of build using colorized output:
+
+`#####green#####>###########black##########> 40% Building`
+
+On ANSI terminals, we can use empty spaces with different background colors for the "bars" and
+A red background shall be used for failed tasks. 
+
+**ASCII-based options (some options use extended ASCII):**
+```
+[##                        ] 6% Building
+[‡‡‡‡                      ] 12% Building
+###########>               > 33% Building
+‹===================       › 60% Building
+«==========================» 100% BUILD SUCCESS
+```
+
+### Implementation
+`DefaultGradleLauncherFactory` registers a listener `BuildProgressBarListener` (similar to `BuildProgressFilter`)
+and forwards events to a `ConsoleBackedProgressBarRenderer`. It would maintain a `BuildProgressBarLogger` that
+logs using a `ProgressBar` implementation of `ProgressFormatter`. A `ProgressBar` has a width and filler character.
+
+`ConsoleBackedProgressBarRenderer` renders the `ProgressBar` using a `ProgressLabel`, which extends `Label` with 
+the following additions:
+ - `void setProgressBar(ProgressBar progressBar)`
+ - `void setText(String text)` (inherited from `Label`)
+ - `void render()` (we can probably add this to `Label`)
+ 
+Under `--continuous` build, the progress bar and multi status text area are reused.
+
+### Test Coverage
+* By default, logs should be streamed with plain output and not throttled when not attached to a Console
+* Color output is output when requested through via the tooling API via `LongRunningOperation.setColorOutput(true)`
+* OutputEvents tooling to API is unaffected except for additional BuildOperations now published
+* Logs should be streamed with plain output when attached Terminal lacks of color or cursor support
+* Renders well on default macOS Terminal, default Windows Console, Cygwin, PowerShell and common Linux Terminals
+
+## Story: Display work-in-progress through ProgressBar
+This adds a visual indicator of what proportion of the build would be complete if all in-progress work is completed.
```

Comment:
But this is about showing how much is currently in progress. My point was that without knowing how long a task will take, this isn't terribly helpful. The yellow bar would basically always have the same size.